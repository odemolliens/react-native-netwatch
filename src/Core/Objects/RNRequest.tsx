import { ILog, LogType } from '../../types';
// @ts-ignore
// BlobFileReader is needed to read the content of the body response (it's type can be blob)
import BlobFileReader from 'react-native/Libraries/Blob/FileReader';

const fromEntries = (arr: any[]) =>
  arr.reduce((acc, [k, v]) => {
    acc[k] = v;
    return acc;
  }, {});

export const stringifyData = (data: any) => {
  try {
    if (data?._parts?.length) {
      return JSON.stringify(fromEntries(data?._parts), null, 2);
    }
    return JSON.stringify(JSON.parse(data), null, 2);
  } catch (e) {
    return `${data}`;
  }
};

export const getRequestBody = (dataSent: any): string => stringifyData(dataSent) || '';

export const getResponseBody = async (responseType: string, response?: any): Promise<string> => {
  if (!response) return '';
  const _responseBody = await (responseType !== 'blob' ? response : parseResponseBlob(response));
  return stringifyData(_responseBody || '');
};

export const parseResponseBlob = async (response: Blob) => {
  const blobReader = new BlobFileReader();
  blobReader.readAsText(response);

  return await new Promise<string>((resolve, reject) => {
    const handleError = () => reject(blobReader.error);

    blobReader.addEventListener('load', () => {
      resolve(blobReader.result);
    });
    blobReader.addEventListener('error', handleError);
    blobReader.addEventListener('abort', handleError);
  });
};

export class RNRequest implements ILog {
  _id: number = -1;
  type: LogType = 'RNR';
  startTime: number = Date.now();
  readyState: number = 0;
  url: string = '';
  shortUrl?: string = '';
  method: string = '';
  status: number = -1;
  endTime: number = 0;
  timeout?: number;
  dataSent: string = '';
  requestHeaders?: any;
  responseHeaders?: any;
  responseContentType?: string;
  responseSize?: number;
  responseType?: string;
  responseURL?: string;
  response: string = '';

  constructor(attributes?: any) {
    Object.assign(this, attributes);
  }
}

export default RNRequest;
