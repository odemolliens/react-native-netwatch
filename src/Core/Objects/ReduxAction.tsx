import { ILog, LogType } from '../../types';

interface IAction {
  type: string;
  payload: any;
}

export class ReduxAction implements ILog {
  _id: number = -1;
  type: LogType = 'REDUX';
  startTime: number = Date.now();
  stringifiedAction: string = '';
  extra: string = '';
  action: IAction = { type: '__ERROR:UNDEFINED__', payload: '' };

  constructor(attributes?: any) {
    Object.assign(this, attributes);
  }
}

export default ReduxAction;
