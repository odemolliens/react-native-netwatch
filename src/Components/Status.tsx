import * as React from 'react';
import { useContext } from 'react';
import { View, StyleSheet } from 'react-native';
import ReduxAction from '../Core/Objects/ReduxAction';
import RNRequest from '../Core/Objects/RNRequest';
import NRequest from '../Core/Objects/NRequest';
import { getStatus } from '../Utils/helpers';
import { ThemeContext } from '../Theme';
import { EnumStatus } from '../types';
import Fontisto from 'react-native-vector-icons/Fontisto';
import { Text } from '../Components/Text';

export interface IProps {
  item: RNRequest | ReduxAction | NRequest;
  text?: string;
  subText?: string;
  textColor?: Object;
  backgroundColor?: string;
}

export const tag = (color: string, text: string) => {
  const theme = useContext(ThemeContext);
  return (
    <View testID="statusCodeColor" style={[styles.status, { backgroundColor: color }]}>
      <Text style={[styles.text, { color: theme.textColorOne }]}>{text}</Text>
    </View>
  );
};

export const reduxTag = () => {
  const theme = useContext(ThemeContext);
  return (
    <View
      testID="statusCodeColor"
      style={[styles.status, { backgroundColor: theme.reduxColor }, { paddingHorizontal: 8 }]}
    >
      <Fontisto name="redux" color={theme.textColorOne} size={12} />
    </View>
  );
};

export const Status: React.FC<IProps> = (props: IProps) => {
  const theme = useContext(ThemeContext);
  let _color: string = theme.textColorOne;
  let _line1: string = '';
  let _line2: string = '';

  if (props.item instanceof ReduxAction) {
    _color = theme.reduxColor;
    _line1 = 'REDUX';
  } else {
    const _temp = getStatus(props.item.status);
    if (_temp === EnumStatus.Success) _color = theme.successColor;
    if (_temp === EnumStatus.Warning) _color = theme.warningColor;
    if (_temp === EnumStatus.Failed) _color = theme.failedColor;
    _line1 = props.item.method;
    _line2 = props.item.status?.toString() || '500';
  }

  return (
    <View style={[styles.container]}>
      <Text>{_line1}</Text>
      {_line2.length > 0 ? tag(_color, _line2) : reduxTag()}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    justifyContent: 'space-evenly',
    alignItems: 'center',
    width: 60,
    height: 60,
    padding: 6,
  },
  status: {
    justifyContent: 'center',
    fontSize: 12,
    height: 22,
    width: 30,
    borderRadius: 4,
  },

  text: {
    textAlign: 'center',
  },
});
