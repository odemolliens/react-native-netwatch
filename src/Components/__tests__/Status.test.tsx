import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { Status, IProps } from '../Status';
import ReduxAction from '../../Core/Objects/ReduxAction';
import RNRequest from '../../Core/Objects/RNRequest';

describe('Status test suite', () => {
  let component: ShallowWrapper;
  let props: IProps;

  it('should render properly redux action', () => {
    givenProps(mockAction);
    givenComponent();
    expect(component).toMatchSnapshot();
  });

  it('should render properly request', () => {
    givenProps(mockRequest);
    givenComponent();
    expect(component).toMatchSnapshot();
  });

  it('should render properly request with method', () => {
    givenProps(mockRequestWithMethod);
    givenComponent();
    expect(component).toMatchSnapshot();
  });

  it('should render properly request with method post and warning', () => {
    givenProps(mockRequestWithMethodPostWarning);
    givenComponent();
    expect(component).toMatchSnapshot();
  });

  it('should render properly request with method post and failed', () => {
    givenProps(mockRequestWithMethodPostFailed);
    givenComponent();
    expect(component).toMatchSnapshot();
  });

  // GIVEN
  function givenComponent() {
    component = shallow(<Status {...props} />);
  }

  function givenProps(item: ReduxAction | RNRequest) {
    props = {
      item,
      text: 'text',
      subText: 'subText',
    };
  }

  const mockAction: ReduxAction = new ReduxAction({
    _id: 73,
    startTime: 100,
    type: 'REDUX',
    action: { type: '__ERROR:UNDEFINED__', action: '' },
  });

  const mockRequest: any = {
    _id: 73,
    dataSent: 'dataSent',
    endTime: 1613477575757,
    readyState: 4,
    response: 'response',
    responseContentType: 'application/json',
    responseSize: 0,
    responseType: 'blob',
    responseURL: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
    startTime: 1613477574742,
    timeout: 0,
    type: 'RNR',
    url: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
  };

  const mockRequestWithMethod: RNRequest = new RNRequest({
    _id: 73,
    dataSent: 'dataSent',
    endTime: 1613477575757,
    method: 'GET',
    readyState: 4,
    response: 'response',
    responseContentType: 'application/json',
    responseSize: 0,
    responseType: 'blob',
    responseURL: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
    startTime: 1613477574742,
    status: 200,
    timeout: 0,
    type: 'RNR',
    url: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
  });

  const mockRequestWithMethodPostWarning: RNRequest = new RNRequest({
    _id: 73,
    dataSent: 'dataSent',
    endTime: 1613477575757,
    method: 'POST',
    readyState: 4,
    response: 'response',
    responseContentType: 'application/json',
    responseSize: 0,
    responseType: 'blob',
    responseURL: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
    startTime: 1613477574742,
    status: 302,
    timeout: 0,
    type: 'RNR',
    url: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
  });
});

const mockRequestWithMethodPostFailed: RNRequest = new RNRequest({
  _id: 73,
  dataSent: 'dataSent',
  endTime: 1613477575757,
  method: 'POST',
  readyState: 4,
  response: 'response',
  responseContentType: 'application/json',
  responseSize: 0,
  responseType: 'blob',
  responseURL: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
  startTime: 1613477574742,
  timeout: 0,
  type: 'RNR',
  url: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
});
