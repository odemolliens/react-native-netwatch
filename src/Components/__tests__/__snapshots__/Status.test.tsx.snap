// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status test suite should render properly redux action 1`] = `
<Component
  style={
    Array [
      Object {
        "alignItems": "center",
        "height": 60,
        "justifyContent": "space-evenly",
        "padding": 6,
        "width": 60,
      },
    ]
  }
>
  <Text>
    REDUX
  </Text>
  <Component
    style={
      Array [
        Object {
          "borderRadius": 4,
          "fontSize": 12,
          "height": 22,
          "justifyContent": "center",
          "width": 30,
        },
        Object {
          "backgroundColor": "#764ABC",
        },
        Object {
          "paddingHorizontal": 8,
        },
      ]
    }
    testID="statusCodeColor"
  >
    <Icon
      color="#F9FAFB"
      name="redux"
      size={12}
    />
  </Component>
</Component>
`;

exports[`Status test suite should render properly request 1`] = `""`;

exports[`Status test suite should render properly request with method 1`] = `
<Component
  style={
    Array [
      Object {
        "alignItems": "center",
        "height": 60,
        "justifyContent": "space-evenly",
        "padding": 6,
        "width": 60,
      },
    ]
  }
>
  <Text>
    GET
  </Text>
  <Component
    style={
      Array [
        Object {
          "borderRadius": 4,
          "fontSize": 12,
          "height": 22,
          "justifyContent": "center",
          "width": 30,
        },
        Object {
          "backgroundColor": "#4D7C0F",
        },
      ]
    }
    testID="statusCodeColor"
  >
    <Text
      style={
        Array [
          Object {
            "textAlign": "center",
          },
          Object {
            "color": "#F9FAFB",
          },
        ]
      }
    >
      200
    </Text>
  </Component>
</Component>
`;

exports[`Status test suite should render properly request with method post and failed 1`] = `
<Component
  style={
    Array [
      Object {
        "alignItems": "center",
        "height": 60,
        "justifyContent": "space-evenly",
        "padding": 6,
        "width": 60,
      },
    ]
  }
>
  <Text>
    POST
  </Text>
  <Component
    style={
      Array [
        Object {
          "borderRadius": 4,
          "fontSize": 12,
          "height": 22,
          "justifyContent": "center",
          "width": 30,
        },
        Object {
          "backgroundColor": "#B91C1C",
        },
      ]
    }
    testID="statusCodeColor"
  >
    <Text
      style={
        Array [
          Object {
            "textAlign": "center",
          },
          Object {
            "color": "#F9FAFB",
          },
        ]
      }
    >
      -1
    </Text>
  </Component>
</Component>
`;

exports[`Status test suite should render properly request with method post and warning 1`] = `
<Component
  style={
    Array [
      Object {
        "alignItems": "center",
        "height": 60,
        "justifyContent": "space-evenly",
        "padding": 6,
        "width": 60,
      },
    ]
  }
>
  <Text>
    POST
  </Text>
  <Component
    style={
      Array [
        Object {
          "borderRadius": 4,
          "fontSize": 12,
          "height": 22,
          "justifyContent": "center",
          "width": 30,
        },
        Object {
          "backgroundColor": "#D97706",
        },
      ]
    }
    testID="statusCodeColor"
  >
    <Text
      style={
        Array [
          Object {
            "textAlign": "center",
          },
          Object {
            "color": "#F9FAFB",
          },
        ]
      }
    >
      302
    </Text>
  </Component>
</Component>
`;
